### PR TITLE
Hackeython/error reporting

### PR DESCRIPTION
**Pull request by mattulbrich**
This brings the edits from the error reporting group at the hackeython onto the main branch

## Intended Change

Have better parser message.

The code driving the parsers has been adapted. Also the code that produces the JML code to parse had to be adapted to contain the correct spaces. Moreover, message extraction from ANTLR exception classes has been improved.

## Type of pull request

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Refactoring (behaviour should not change or only minimally change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] There are changes to the (Java) code
- [ ] There are changes to the taclet rule base
- [ ] There are changes to the deployment/CI infrastructure (gradle, github, ...)
- [ ] Other: 

## Ensuring quality
    
- [x] I made sure that introduced/changed code is well documented (javadoc and inline comments).
- [ ] I made sure that new/changed end-user features are well documented (https://github.com/KeYProject/key-docs).
- [x] I added new test case(s) for new functionality.
- [ ] I have tested the feature as follows: ...
- [ ] I have checked that runtime performance has not deteriorated.

## Additional information and contact(s)

Collaboration with @FliegendeWurst @MarcoScaletta @LennartKleinwort

The contributions within this pull request are licensed under GPLv2 (only) for inclusion in KeY.

### CODE REVIEW

**Comment by FliegendeWurst** [comment]
This PR will also fix #3441

**Comment by mattulbrich** [comment]
Ping for a review, @WolframPfeifer. Thx.

File: key.core/src/test/resources/de/uka/ilkd/key/nparser/exceptional/doubleDecl2.key [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+
+\programVariables {
+    int null;
+}
+
+\problem {true}
+
+
+// This crashes everything :(
```
Comment (WolframPfeifer) [inline key.core/src/test/resources/de/uka/ilkd/key/nparser/exceptional/doubleDecl2.key:15]: Should this have `broken: true` then? Or is this comment outdated?

File: key.core/src/test/resources/de/uka/ilkd/key/speclang/njml/exceptional/Bevhavioural.java [hunk 2]
```diff
@@ -1,7 +1,7 @@
 // exceptionClass: PosConvertException
 // msgContains: no viable alternative at input 'normal_bevhaviour
 // verbose: true
-// broken: true
+// broken: false
 
 // This is broken since it currently reports the wrong location
```
Comment (WolframPfeifer) [inline key.core/src/test/resources/de/uka/ilkd/key/speclang/njml/exceptional/Bevhavioural.java:6]: Inconsistency between comment and `broken` flag ...

File: key.core/src/test/resources/de/uka/ilkd/key/speclang/njml/exceptional/NamedAccessInContract.java [hunk 3]
```diff
@@ -1,8 +1,7 @@
 // position: 12/20
 // verbose: true
 // broken: true
-// exceptionClass: XXXX
-// msgContains: XXXX
+// noException: false
 
 // currently does not report an exception
```
Comment (WolframPfeifer) [inline key.core/src/test/resources/de/uka/ilkd/key/speclang/njml/exceptional/NamedAccessInContract.java:6]: Looks like another inconsistency between comment and `noException` tag.

**Review by WolframPfeifer — APPROVED** [review]
I think this PR in general enhances the error messages. While I did not check all the parser message test cases, I think the fact that there is an infrastructure to ensure that in certain cases specific messages are thrown is an improvement.

The PR is also nicely documented, however, I pointed out a few places where comments and the tags are inconsistent. However, as this is no functional problem and the PR is already waiting very long, I think we can merge this now and update the 3 comments later.

**Pull request merged**